packages/react/src/components/Tabs/Tabs.tsx

```diff
@@ -0,0 +1,157 @@
+  variant,
+  className
+}: TabsProps): JSX.Element => {
+  const [activeIndex, setActiveIndex] = useState(initialActiveIndex);
+  const [id] = propId ? [propId] : useId(1, 'tabs');
+  const focusedTabRef = useRef<HTMLLIElement>(null);
+
+  const tabs = React.Children.toArray(children).filter(
+    child => (child as React.ReactElement<any>).type === Tab
+  );
+  const panels = React.Children.toArray(children).filter(
+    child => (child as React.ReactElement<any>).type === TabPanel
+  );
+
+  const tabCount = tabs.length;
+
+  const handleClick = (index: number) => {
+    setActiveIndex(index);
+  };
+
+  const handleKeyDown = (event: React.KeyboardEvent) => {
+    const { which } = event;
+    let newIndex: number = activeIndex;
+
+    switch (which) {
+      case left: {
+        newIndex = activeIndex - 1;
+
+        // circularity
+        if (newIndex === -1) {
+          newIndex = tabCount - 1;
+        }
+        setActiveIndex(newIndex);
+        break;
+      }
+      case right: {
+        newIndex = activeIndex + 1;
+
+        // circularity
+        if (newIndex === tabCount) {
+          newIndex = 0;
+        }
+        setActiveIndex(newIndex);
+        break;
+      }
+      case home: {
+        newIndex = 0;
+        setActiveIndex(newIndex);
+        break;
+      }
+      case end: {
+        newIndex = tabCount - 1;
+        setActiveIndex(newIndex);
+        break;
+      }
+    }
+  };
+
+  const tabComponents = tabs.map((child, index) => {
+    const { ...other } = (child as React.ReactElement<any>).props;
+    const selected = index === activeIndex;
+
+    const config = {
+      id: `${id}-${index}`,
+      className: classNames('Tab', {
+        'Tab--active': selected,
+        'Tab--full-width': variant === 'full-width'
+      }),
+      tabIndex: index === activeIndex ? 0 : -1,
+      ['aria-controls']: `${id}-panel-${index}`,
+      ['aria-selected']: selected,
+      ref: index === activeIndex ? focusedTabRef : null,
+      onClick: () => handleClick(index),
+      ...other
+    };
+
+    return React.cloneElement(child as React.ReactElement<any>, config);
+  });
+
+  const tabPanelComponents = panels.map((child, index) => {
+    const { className, ...other } = (child as React.ReactElement<any>).props;
+    const panelId = `${id}-panel-${index}`;
+    return React.cloneElement(child as React.ReactElement<any>, {
+      id: panelId,
+      ['aria-labelledby']: `${id}-${index}`,
+      className: classNames('TabPanel', className, {
+        'TabPanel--hidden': activeIndex !== index
+      }),
+      ...other
+    });
+  });
+
+  useDidUpdate(() => {
+    focusedTabRef.current?.focus();
+  }, [activeIndex]);
+
+  return (
+    <div
+      className={classNames('Tabs', className, {
+        'Tabs--thin': thin
+      })}
+    >
+      <ul
+        role="tablist"
+        className={classNames('Tablist', {
+          'Tablist--full-width': variant === 'full-width'
+        })}
+        aria-label={label}
```

Comment:
I hate to keep going back and forth on this... @scurker and I talked about this yesterday but there may have been a misunderstanding...I don't think we should force an `aria-label` on the tablist because `aria-labelledby` is just as good, if not better in some situations. imo we should force the caller to either provide an `aria-label` prop or an `aria-labelledby` prop

Reply:
To expand on my rationale a bit more - 

Often times tabs will appear directly beneath a heading and in those situations it is best to just make the tablist aria-labelledby the heading like so:


```js
<h3 id="foo">Stuff/things/etc</h3>
<div role="tablist" aria-labelledby="foo">...</div>
```

Reply:
I think @schne324 is right here, and I overlooked that.

You can add an intersecting type allowing `aria-labelledby` or `aria-label`:

```ts
type props = {
  foo: 'string'
} & ({ labelledby: string } | { label: string })

type LabelProps = { 'aria-label': string } | { 'aria-labelledby': string }
type Props = {
  foo: 'bar'
} & LabelProps

// fails because it does not have aria-label or aria-labelledby
const props1: Props = { 
  foo: 'bar'
}

// These pass because they satisfy one of the union conditions
const props2: Props = {
  foo: 'bar',
  'aria-label': 'hello'
}
const props3: Props = {
  foo: 'bar',
  'aria-labelledby': 'hello'
}
```